app/server.ts

```diff
@@ -1,33 +1,59 @@
+import crypto from "crypto";
 import { createServer } from "node:http";
 import next from "next";
 import { Server } from "socket.io";
-import type { ClientToServerEvents, ServerToClientEvents } from "@/types";
+import type { SocketServer } from "@/types";
 import { Field } from "@/utils/game";
 
 const dev = process.env.NODE_ENV !== "production";
 const hostname = "localhost";
 const port = 3000;
+const sessions = new Map<string, "alive" | "dead">();
 
 async function main() {
   const app = next({ dev, hostname, port });
   await app.prepare();
 
   const httpServer = createServer(app.getRequestHandler());
-  const io = new Server<ClientToServerEvents, ServerToClientEvents>(httpServer);
+  const io: SocketServer = new Server(httpServer);
 
   let clientsCount = 0;
 
   let field = await Field.fromRedis();
   field = await field.handleComplete();
 
+  io.use(async (socket, next) => {
+    const { sessionID } = socket.handshake.auth;
+    if (sessionID && sessions.has(sessionID)) {
+      if (sessions.get(sessionID) === "dead") {
+        return next(new Error("dead"));
+      }
+      socket.data.sessionID = sessionID;
+      return next();
+    }
+
+    socket.data.sessionID = crypto.randomBytes(8).toString("hex");
+    next();
+  });
+
   io.on("connection", (socket) => {
+    sessions.set(socket.data.sessionID, "alive");
+
+    socket.emit("session", {
+      sessionID: socket.data.sessionID,
+    });
+
     clientsCount++;
     io.emit("clientsCount", clientsCount);
     io.emit("exposedPercent", field.exposedPercent);
     socket.emit("update", field.plots);
 
     socket.on("expose", async (index) => {
-      field.exposeCell(index);
+      const newSessionValue = field.exposeCell(index);
+      if (newSessionValue === "dead") {
+        sessions.set(socket.data.sessionID, "dead");
+        socket.emit("death");
+      }
       field = await field.handleComplete();
       io.emit("exposedPercent", field.exposedPercent);
       io.emit("update", field.plots);
```

Comment:
we can omit naming this variable (I heard you don't like naming things) also, we may not have to do anything other than emit `"update"` afterwards
```suggestion
      if (field.exposeCell(index) === "dead") {
        sessions.set(socket.data.sessionId, "dead");
        socket.emit("death");
      } else {
        field = await field.handleComplete();
        io.emit("exposedPercent", field.exposedPercent);
      }
      io.emit("update", field.plots);
```